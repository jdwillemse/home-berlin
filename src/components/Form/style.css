.input__wrap {
  display: block;
  margin: 0 0 20px;
  padding: 0;
  border: none;
}

.input__label {
  /* float: left; */
  /* width: 30%; */
}

.input__field {
  width: 100%;
}

.input__radio {
  display: block;
}

.input__field--radio,
.input__label--radio {
  float: none;
  width: auto;
}

.progress-indicator {
  overflow: hidden;
  margin-bottom: 20px;
  width: 100%;
}

.progress-indicator__inner {
  width: 100%;
  height: 5px;
  transition: transform 0.3s ease-in-out;
  transform: translate3d(-100%, 0, 0);
  background: #0077ff;
}

.progress-indicator__inner--25 {
  transform: translate3d(-75%, 0, 0);
}

.progress-indicator__inner--50 {
  transform: translate3d(-50%, 0, 0);
}

.progress-indicator__inner--75 {
  transform: translate3d(-25%, 0, 0);
}

.progress-indicator__inner--100 {
  transform: translate3d(0, 0, 0);
}

@media (min-width: 500px) {
  .input__wrap {
    margin-bottom: 40px;
  }

  .input__label {
    float: left;
    width: 35%;
  }

  .input__field {
    width: 65%;
  }

  .input__radio {
    display: inline-block;
    width: 50%;
  }

  .input__field--radio,
  .input__label--radio {
    float: none;
    width: auto;
  }
}
