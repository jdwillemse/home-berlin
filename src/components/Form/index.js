import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import './style.css';
import Input from '../Input';

const salaryBrackets = {
  bracket1: 'Less than 1.000',
  bracket2: '1.000 – 2.000',
  bracket3: '2.000 – 3.000',
  bracket4: '3.000 – 4.000',
  bigMoney: 'More than 4.000'
};

/**
 * A custom Form component that handles form validation errors.
 * It executes the form's checkValidity
 */
class Form extends React.PureComponent {
  /**
   * Define local component state in class constructor.
   * @param {object} props - Initial compoenent props
   */
  constructor(props) {
    super(props);

    this.state = {
      percentProgress: 0
    };

    this.fieldsets = [];
    this.fieldsetCount = 4;
    this.formData = [];

    this.addRef = this.addRef.bind(this);
    this.measureProgress = this.measureProgress.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  addRef(el) {
    this.fieldsets.push(el);
  }

  /**
   * The main function that validates the form and fills in the error messages.
   * Based on https://codepen.io/_arpy/pen/xYoyPW
   * @return {bool} - Returns a boolean showing if the form is valid for submission or not.
   */
  validate() {
    // this.formEl is a reference in the component to the form DOM element.
    const { formEl } = this;
    const formLength = this.fieldsets.length;

    /*
     * The checkValidity() method on a form runs the
     * html5 form validation of its elements and returns the result as a boolean.
     * It returns 'false' if at least one of the form elements does not qualify,
     * and 'true', if all form elements are filled with valid values.
     */
    if (formEl.checkValidity() === false) {
      for (let i = 0; i < formLength; i++) {
        const fieldset = this.fieldsets[i];
        const elem = fieldset.querySelector('input');
        const errorLabel = fieldset.querySelector('.invalid-feedback');

        if (errorLabel) {
          /*
           * Each node in html5 form has a validity property.
           * It contains the validation state of that element.
           * The elem.validity.valid property indicates whether the element qualifies its validation rules or no.
           * If it does not qualify, the elem.validationMessage property will contain the localized validation error message.
           * We will show that message in our error container if the element is invalid, and clear the previous message, if it is valid.
           */
          if (!elem.validity.valid) {
            errorLabel.textContent = elem.validationMessage;
            errorLabel.setAttribute('aria-hidden', 'false');
          } else if (errorLabel.textContent) {
            errorLabel.textContent = '';
            errorLabel.setAttribute('aria-hidden', 'true');
          }
        }
      }

      // Return 'false', as the formEl.checkValidity() method found some invalid form inputs.
      return false;
    }

    // The form is valid, so we clear all the error messages
    for (let i = 0; i < formLength; i++) {
      const errorLabel = this.fieldsets[i].querySelector('.invalid-feedback');
      if (errorLabel) {
        errorLabel.textContent = '';
      }
    }

    // Return 'true', as the form is valid for submission
    return true;
  }

  /**
   * Check form completion progress by testing if input fields have values. Validity is not taken into
   * account at this point, only completedness.
   * @param {object} - form input element
   * @return {undefined}
   */
  measureProgress({ target: { validity, name, value, type } }) {
    let filledFieldsCount = 0;
    // strip out spacing characters from phone number
    this.formData[name] = type === 'tel' ? value.replace(/-|\s/g, '') : value;

    filledFieldsCount = Object.values(this.formData).filter(input => input.trim()).length;

    this.setState({
      percentProgress: (filledFieldsCount / this.fieldsetCount) * 100
    });
  }

  /**
   * Pass form data to next route as state.
   * It stops the default form submission process and proceeds with custom validation.
   * @param {object} event - Synthetic submit event
   * @return {undefined}
   */
  handleSubmit(event) {
    const { history } = this.props;

    event.preventDefault();

    if (this.validate()) {
      // passing form data as router state is the easiest way to send it to the summary page;
      // in a more complex app I would have used redux
      history.push({ pathname: '/summary', state: this.formData });
    }
  }

  /**
   * Render the component.
   * @return {node} - JSX node
   */
  render() {
    const { percentProgress } = this.state;
    const inputProps = {
      onBlur: this.measureProgress,
      addRef: this.addRef
    };

    return (
      <form
        noValidate
        ref={form => {
          this.formEl = form;
        }}
        onSubmit={this.handleSubmit}
      >
        <div className="progress-indicator">
          <div
            className={`progress-indicator__inner progress-indicator__inner--${percentProgress}`}
            aria-label={`form is ${percentProgress}% complete`}
            aria-live="polite"
          />
        </div>

        <Input name="fullname" label="Full name" type="text" {...inputProps} />
        <Input name="email" label="Email address" type="email" {...inputProps} />
        <Input
          name="phone"
          label="Phone number"
          type="tel"
          pattern="^\+?(\d|-|\s)+$"
          {...inputProps}
        />

        <fieldset className="input__wrap clearfix" ref={this.addRef}>
          {Object.keys(salaryBrackets).map((key, i) => (
            <label htmlFor={key} key={key} className="input__radio">
              <input
                type="radio"
                name="salary"
                id={key}
                value={key}
                className="input__field input__field--radio"
                aria-describedby="salary-error"
                onChange={this.measureProgress}
                required
              />
              <span className="input__label input__label--radio">{salaryBrackets[key]}</span>
            </label>
          ))}
          <div className="invalid-feedback" id="salary-error" aria-live="polite" />
        </fieldset>

        <button type="submit">submit</button>
      </form>
    );
  }
}

Form.propTypes = {
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired
};

export default withRouter(Form);
