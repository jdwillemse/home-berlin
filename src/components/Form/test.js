import React from 'react';

import Form from './index';

describe('Form', () => {
  beforeAll(() => {
    // Mock warn & error to catch errors thrown by React into the console
    console.warn = jest.fn();
    console.error = jest.fn();
  });

  afterEach(() => {
    expect(console.warn).not.toBeCalled();
    expect(console.error).not.toBeCalled();
  });

  it('should match snapshot', () => {
    const wrapper = shallow(<Form history={{ push: jest.fn() }} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
