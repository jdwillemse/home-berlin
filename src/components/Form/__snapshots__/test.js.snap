// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="progress-indicator"
  >
    <div
      aria-label="form is 0% complete"
      aria-live="polite"
      className="progress-indicator__inner progress-indicator__inner--0"
    />
  </div>
  <Input
    addRef={[Function]}
    label="Full name"
    name="fullname"
    onBlur={[Function]}
    type="text"
  />
  <Input
    addRef={[Function]}
    label="Email address"
    name="email"
    onBlur={[Function]}
    type="email"
  />
  <Input
    addRef={[Function]}
    label="Phone number"
    name="phone"
    onBlur={[Function]}
    pattern="^\\\\+?(\\\\d|-|\\\\s)+$"
    type="tel"
  />
  <fieldset
    className="input__wrap clearfix"
  >
    <label
      className="input__radio"
      htmlFor="bracket1"
      key="bracket1"
    >
      <input
        aria-describedby="salary-error"
        className="input__field input__field--radio"
        id="bracket1"
        name="salary"
        onChange={[Function]}
        required={true}
        type="radio"
        value="bracket1"
      />
      <span
        className="input__label input__label--radio"
      >
        Less than 1.000
      </span>
    </label>
    <label
      className="input__radio"
      htmlFor="bracket2"
      key="bracket2"
    >
      <input
        aria-describedby="salary-error"
        className="input__field input__field--radio"
        id="bracket2"
        name="salary"
        onChange={[Function]}
        required={true}
        type="radio"
        value="bracket2"
      />
      <span
        className="input__label input__label--radio"
      >
        1.000 – 2.000
      </span>
    </label>
    <label
      className="input__radio"
      htmlFor="bracket3"
      key="bracket3"
    >
      <input
        aria-describedby="salary-error"
        className="input__field input__field--radio"
        id="bracket3"
        name="salary"
        onChange={[Function]}
        required={true}
        type="radio"
        value="bracket3"
      />
      <span
        className="input__label input__label--radio"
      >
        2.000 – 3.000
      </span>
    </label>
    <label
      className="input__radio"
      htmlFor="bracket4"
      key="bracket4"
    >
      <input
        aria-describedby="salary-error"
        className="input__field input__field--radio"
        id="bracket4"
        name="salary"
        onChange={[Function]}
        required={true}
        type="radio"
        value="bracket4"
      />
      <span
        className="input__label input__label--radio"
      >
        3.000 – 4.000
      </span>
    </label>
    <label
      className="input__radio"
      htmlFor="bigMoney"
      key="bigMoney"
    >
      <input
        aria-describedby="salary-error"
        className="input__field input__field--radio"
        id="bigMoney"
        name="salary"
        onChange={[Function]}
        required={true}
        type="radio"
        value="bigMoney"
      />
      <span
        className="input__label input__label--radio"
      >
        More than 4.000
      </span>
    </label>
    <div
      aria-live="polite"
      className="invalid-feedback"
      id="salary-error"
    />
  </fieldset>
  <button
    type="submit"
  >
    submit
  </button>
</form>
`;
